ratelimiter/ratelimiter.go

```diff
@@ -0,0 +1,147 @@
+package ratelimiter
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/juju/ratelimit"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+// Interface describes common ratelimiter methods.
+type Interface interface {
+	Create([]byte, Config) error
+	Remove([]byte, time.Duration) error
+	TakeAvailable([]byte, int64) int64
+	Available([]byte) int64
+}
+
+// Config is a set of options used by rate limiter.
+type Config struct {
+	Interval, Capacity, Quantum uint64
+}
+
+// compare config with existing ratelimited bucket.
+func compare(c Config, bucket *ratelimit.Bucket) bool {
+	return int64(c.Capacity) == bucket.Capacity() &&
+		1e9*float64(c.Quantum)/float64(c.Interval) == bucket.Rate()
+}
+
+func newBucket(c Config) *ratelimit.Bucket {
+	return ratelimit.NewBucketWithQuantum(time.Duration(c.Interval), int64(c.Capacity), int64(c.Quantum))
+}
+
+// NewPersisted returns instance of rate limiter with persisted black listed records and capacity before peer was removed.
+func NewPersisted(db DBInterface) *PersistedRateLimiter {
+	return &PersistedRateLimiter{
+		db:          db,
+		initialized: map[string]*ratelimit.Bucket{},
+		timeFunc:    time.Now,
+	}
+}
+
+// PersistedRateLimiter persists latest capacity and updated config per unique ID.
+type PersistedRateLimiter struct {
+	db DBInterface
+
+	mu          sync.Mutex
+	initialized map[string]*ratelimit.Bucket
+
+	timeFunc func() time.Time
+}
+
+func (r *PersistedRateLimiter) blacklist(id []byte, duration time.Duration) error {
+	if duration == 0 {
+		return nil
+	}
+	record := BlacklistRecord{ID: id, Deadline: r.timeFunc().Add(duration)}
+	if err := record.Write(r.db); err != nil {
+		return fmt.Errorf("error blacklisting %x: %v", id, err)
+	}
+	return nil
+}
+
+func (r *PersistedRateLimiter) get(id []byte) (bucket *ratelimit.Bucket, exist bool) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	bucket, exist = r.initialized[string(id)]
+	return
+}
+
+// Create creates an instance for a provided ID. If ID was blacklisted error is returned.
+func (r *PersistedRateLimiter) Create(id []byte, cfg Config) error {
+	bl := BlacklistRecord{ID: id}
```

Comment:
can we make something like `isBlacklisted()` function from this? it took some efforts to understand this piece and what it has to do with creating.

```go
func (r *PersistedRateLimiter) Create(id []byte, cfg Config) error {
    if r.isBlacklisted() {
      return err...
    }
}
```
should be more clear.